docs/design/coreclr/botr/data_contracts_for_coreclr.md

```diff
@@ -0,0 +1,430 @@
+# Data Contracts for CoreCLR, and the data contract stream
+
+Data contracts represent the manner in which a tool which is not the runtime can reliably understand and observe the behavior of the runtime. Contracts are defined by their documentation, and the runtime describes what contracts are applicable to understanding that runtime.
+
+## Data Contract Stream
+The physical layout of this stream is not defined in this document, but its practical effects are.
+
+The data contract stream has a set of records of the following forms.
+
+### Global Values
+Global values which can be of types (int8, uint8, int16, uint16, int32, uint32, int64, uint64, pointer, unicode string)
+All global values have a unicode string describing their name, and a value of one of the above types.
+
+### Compatible Contract
+Each compatible contract is described by a unicode string naming the contract, and a uint32 version. It is an ERROR if multiple versions of a contract are specified in the stream.
+
+### Data Structure Layout
+Each data structure layout has a unicode name for the type, followed by a list of fields. These fields can be of primitive types (int8, uint8, int16, uint16, int32, uint32, int64, uint64, pointer) or of another named data structure type. Each field descriptor provides the offset of the field, the name of the field, and the type of the field.
+
+## Versioning of contracts
+Contracts are described an integer version number. A higher version number is not more recent, it just means different. In order to avoid conflicts, all contracts should be documented in the main branch of the dotnet repository with a version number which does not conflict with any other. It is expected that every version of every contract describes the same functionality/data layout/set of global values.
```

Comment:
Can we get a concrete example of how this would work? I don't see the need for a minor version number.

Reply:
I partly agree. I think it's important that a data contract reader is able to reliably tell when it is talking to a runtime that is newer than itself and has a reasonable behavior (either detect that it is out of date or fall back to some compatible behavior).

Mono debugger-libs takes the "compatible behavior" route.  the protocol is major/minor versioned and each operation does an "target version is at least X.Y" check when it needs to decide whether to use a newer or older version of an algorithm (or if it needs to detect that the wire data has added fields).

---

one approach might be to say that the data contract reader knows every contract+version that it is expected to support.  if the descriptor includes a contract+version that isn't supported, the reader should abort.

additionally each contract spec should include a list of "subsumed" versions.  So contract MethodTable,5 can say it subsumes versions 4,3,1.  Whereas MethodTable,2 is unrelated to 5 (either because it is a breaking change or because it's for a different runtime).  So a reader implementation for MethodTable,5 can also understand runtimes that expose MethodTable,4 or MethodTable,3.  Whereas presumably MethodTable,2 needs an entirely separate implementation.

Reply:
It expect that it is not going to be unusual to amend existing contracts in non-breaking way. It is important to be able to do this in servicing where you do not want to break existing consumers, but you want consumers aware of the amendment to be able to detect it and take advantage of it.

In the scheme with major versions only, you can achieve this by exposing the amendment of `MyContract` as a new `MyContractEx`. 

In the scheme with major.minor versions, you can achieve this by incrementing the minor version of `MyContract` and include the amendment in `MyContract`.

Reply:
That also is acheived by optional fields and the processing of the data understanding when data is optional vs when it is a required. This notion is captured in data being able to describe itself and the fields it provides. Contract version X indicates fields a and b. If more data is needed, then it can add field c. The processing only knows about a and b and so the new optional field is a non-breaking change. When version X+1 is created, field c can become required or remain optional. I think how one writes the algorithm should dictate the breaking changes, not the data itself.

Reply:
> acheived by optional fields 

Yes, this works for data structure contracts. For algorithmic contracts, you can achieve this by introducing new key/value pair that indicate the newer version.

(The same alternative representations would work for major versions too. The major versions are not strictly required either.)

Reply:
>  For algorithmic contracts, you can achieve this by introducing new key/value pair that indicate the newer version.

I'm still dubious of the independent versioning of algorithmic contracts. My biggest concern here is the idea that that design creates an inherent maintenance issue. The issue being we have versioned data and versioned algorithms, so how do they map? If I update Data version X, do I need to update Algorithm version Y? Answering that question should not require a lot of system understanding. It should be defined by the algorithm in some explicit manner.

One could imagine that at reader start-up, or lazily, an audit of the data stream is performed - see the Fall prototype. Each algorithm is then asked to see what is supports and queries all data contracts provided by the target debuggee. If the algorithm see the versions of data contracts that satisfy it, then the algorithm can do its job, otherwise it can't. This reduces any congnitive load for how to version anything other than algorithms as they are versioned based on the data they consume.

Reply:
> The issue being we have versioned data and versioned algorithms, so how do they map? If I update Data version X, do I need to update Algorithm version Y?

Data structure layouts do not have versions in the current proposals. You use Algorithm version Y with whatever data structure layout is provided.

I general, I agree that the data and algorithms that interpret the data are frequently coupled together, and it is not obvious why they are treated as two different concepts. @noahfalk commented on it as well.

Reply:
> That also is acheived by optional fields and the processing of the data understanding when data is optional vs when it is a required.

I agree that you can do non-breaking changes via optional additive data elsewhere and if we want to define that this is the way all non-breaking changes will be introduced I'm fine with it. My high order bit was that the design should anticipate making non-breaking changes and describe a mechanism we will use to support it.

Reply:
The detail I was getting at with separating layouts from the algorithmic contracts is that while *some* data layouts are implicit in the algorithms, many of them are just working with fields in some set of structures, and on various architectures/os/whatever, the actual layouts of fields in memory can vary a fair bit. Thus the algorithms can be written on top of a fields concept, and the data structure layout describes how to map those fields to actual memory.

Now, some algorithms also rigidly describe the associated data layouts. In those cases there would likely be no reason to have a data structure layout. For instance, the layout of the unwind data for x64 windows unwinding doesn't need a separate layout, the exact layout is well known and part of the algorithm.